mobilecoind/src/payments.rs

```diff
@@ -1078,62 +1232,91 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
+                .add_output(outlay.amount, &outlay.receiver, rng)
                 .map_err(|err| Error::TxBuild(format!("failed adding output: {err}")))?;
 
             tx_out_to_outlay_index.insert(tx_out, i);
             outlay_confirmation_numbers.push(confirmation);
 
-            total_value += outlay.value;
+            *balance_sheet.entry(outlay.amount.token_id).or_default() +=
+                outlay.amount.value as i128;
         }
 
-        // Figure out if we have change.
-        let input_value = inputs
-            .iter()
-            .fold(0, |acc, (utxo, _proof)| acc + utxo.value);
-        if total_value > input_value {
-            return Err(Error::InsufficientFunds);
-        }
-        let change = input_value - total_value - tx_builder.get_fee();
-
-        // If we do have nonzero change, add an output for that as well.
-        // TODO (GH #1522): Should the exchange write destination memos?
-        // If so then we must always write a change output, even if the change is zero
-        if change > 0 {
-            // TODO: If you want to support mixed transactions, use outlay-specific token id
-            // here
-            let change_amount = Amount {
-                value: change,
-                token_id,
-            };
-
-            let change_dest = ReservedSubaddresses::from_subaddress_index(
-                from_account_key,
-                Some(change_subaddress),
-                None,
-            );
+        // Figure out if we have change. Change occurs when the total value of the
+        // inputs exceeds the value of the outlays, so we have a negative entry
+        // in the balance sheet.
+        let change_dest = ReservedSubaddresses::from_subaddress_index(
+            from_account_key,
+            Some(change_subaddress),
+            None,
+        );
 
-            tx_builder
-                .add_change_output(change_amount, &change_dest, rng)
-                .map_err(|err| Error::TxBuild(format!("failed adding output (change): {err}")))?;
+        for (token_id, val) in balance_sheet.iter() {
+            if *val > 0 {
+                log::error!(
+                    logger,
+                    "After input selection we still had an insufficient amount of {}: {}",
+                    token_id,
+                    val
+                );
+                return Err(Error::InsufficientFunds);
+            }
+            // Note: RTH normally requires that we write change even if it's zero, but we
+            // haven't done that here yet, and we're always writing empty memos.
+            // Note: RTH doesn't tolerate having multiple change outputs right now IIRC,
+            // so we would have to extend the scope of RTH to work with complex mixed
+            // transactions that have multiple change outputs, if we want that
+            // to work.
+            if *val < 0 {
+                let change_amount = Amount::new(-*val as u64, *token_id);
```

Comment:
Do we need to make sure the change fits in a u64? I know the worst case is the tx wont get accepted, but it might make debugging easier if we check for that.